Bugfix/rusu1/visit 4632 plot3d iblanking support

rusu24edward opened the pull request:
### Description

Resolves #4632 

There are two changes made here:
1. Added iblanking support to plot3d reader so the reader will now read iblanking data if applicable.
2. Changed how "if applicable" is determined for iblanking.

PLOT3D options are determined like so:
   * If binary file, use auto-detection.
   * Elseif vp3d, use options in vp3d.
   * Else use options in reader options.

It still mostly works this way, but iblanking is an exception here. For iblanking, now we do:
   * If IBlanking is set in vp3d, then use that value
   * Elseif IBlanking is set in reader options, then use that value
   * Else use auto detection.
      * If the file is ascii, then auto-detect won't work and we'll use the default values. The default for iblank is auto-detect, which is a contradiction when the file is ascii. So the default becomes false for ascii files.

This allows the user to manually override the auto-detection for IBlanking. I needed to do this in order to pass regression tests.

### Type of change
Bugfix.

### How Has This Been Tested?
User provided test data that showcased the bug. I applied the fix to this dataset and verified with user that it worked as expected.

I built the docs with the changes and confirmed that [they appear as expected](https://visit-sphinx-github-user-manual.readthedocs.io/en/bugfix-rusu1-visit-4632-plot3d-iblanking-support/gui_manual/WorkingWithFiles/Read_Write_Options.html#plot3d).

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [X] I have updated the release notes
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added any new baselines to the repo
- [x] I have assigned reviewers (see [VisIt's PR procedures](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers) for more information).

rusu24edward committed:
Copied over some code from vtk reader. Still need to fill the iblanking array with info from the file.

rusu24edward committed:
using avt ghost zone types instead

rusu24edward committed:
Trying to identify iblank marked nodes and use avt ghost node type

rusu24edward committed:
Attempting to fill iblank array with iblanking data from file

rusu24edward committed:
Tried a few different things to get it working, but still bad EOF error

rusu24edward committed:
Little more work getting a bit closer to solution

rusu24edward committed:
iblanked nodes and zones are detected correctly

rusu24edward committed:
Added function to process the iblanking read for both binary and nonbinary style

rusu24edward committed:
Cleaned up the code

rusu24edward committed:
Updated test script to test for iblanking fix

rusu24edward committed:
Added iblanking data to PLOT3d

rusu24edward committed:
Migrated visit users docs on plot3d to sphinx

rusu24edward committed:
Updated release notes

rusu24edward committed:
Trying to get codeblock to work in rst

rusu24edward committed:
Try again with rst code block

rusu24edward committed:
Added clarifying comment in reader

rusu24edward committed:
Another attempt for code blocks in docs

rusu24edward committed:
Trying yet again for the code blocks in rst

rusu24edward committed:
Added new baselines

rusu24edward commented:
TODO:
- [x] Figure out why existing tests are failing. Here's an example:
![Screen Shot 2020-11-02 at 8 07 16 PM](https://user-images.githubusercontent.com/18565891/97949011-18bf3600-1d47-11eb-8704-4c19962bc84d.png)

CheckBlankingAndPrecision might be the answer. For binary type, the file itself is checked for blanking. If it determines true, then this overwrites the user's settings in the reader options and in the vp3d file. This seems like a bad user experience: The user can explicitly say NO_BLANKING and VisIt will ignore this and do the blanking because this check determines that based on the file size... Perhaps this check should only happen if the user doesn't specify any settings for IBLANKING in the vp3d file. Reader options can be updated to be auto, yes, no, where this check only happens on auto.

rusu24edward commented:
@biagas This is still in draft mode, but if you see me going down the wrong path or have any insights to provide, please let me know!

biagas commented on src/databases/PLOT3D/vtkPLOT3DReader.C:
```diff
@@ -790,20 +823,77 @@ vtkPLOT3DReader::ReadGrid(FILE *xyzFp)
+    iblank->SetName("avtGhostNodes");
+    iblank->SetNumberOfTuples(this->NumberOfPoints);
+
+    // Attempt to read the iblank data from the plot3d file. If the read failes,
```
minor typo 'failes'

biagas commented on src/test/tests/databases/PLOT3D.py:
```diff
@@ -178,5 +179,55 @@ def BasicTests():
 TestTimeSeries()
-    
+
+# Custom tests with specific views:
```
While there isn't really a rule for this, it is helpful to have different sections of a test file in their own functions. This makes it easier in the future to turn off aspects of the test file when trying to debug. (one # comment to turn off a function call rather than having to comment out a bunch of lines of code. )

biagas commented on src/doc/gui_manual/WorkingWithFiles/Read_Write_Options.rst:
```diff
@@ -110,6 +110,169 @@ all information related to the material configuration during the *open*. This wi
+  2D
+  # 3D
+  
+  # Precision.  Single precision is assumed unless stated otherwise
```
Add period to end.

biagas commented on src/doc/gui_manual/WorkingWithFiles/Read_Write_Options.rst:
```diff
@@ -110,6 +110,169 @@ all information related to the material configuration during the *open*. This wi
+  # STRUCTURED
+  # UNSTRUCTURED
+  
+  # Iblanking.  No iblanking is assumed unless stated otherwise
```
Add period to end.

biagas commented on src/doc/gui_manual/WorkingWithFiles/Read_Write_Options.rst:
```diff
@@ -110,6 +110,169 @@ all information related to the material configuration during the *open*. This wi
+  MULTI_GRID
+  # SINGLE_GRID
+  
+  # Data encoding, ASCII or Binary  Binary will be assumed if no value is
```
Add period between Binary and Binary.

biagas commented on src/doc/gui_manual/WorkingWithFiles/Read_Write_Options.rst:
```diff
@@ -110,6 +110,169 @@ all information related to the material configuration during the *open*. This wi
 
+PLOT3D
+~~~~~~
+
```
Great job getting the documentation in!

biagas reviewed (changes requested):
I think you have done a great job, there are just a few minor things to address.

biagas commented:
@rusu24edward regarding current tests failing, are any of them attempting to go the IBlanking route? If so, to get current results, it will have to be explicitly turned off for them.

I like your idea of a 3-choice option for IBlanking ( auto-on-off ).

biagas commented:
Thinking more about the three-choice, it doesn't really make sense to have an 'on' option if there is no IBlanking in the file. Perhaps it is sufficient to have a way to for user to turn it off.  Just one option: IBLANKING_OFF, if set the reader ignores IBlanking, otherwise it performs a check and uses it if present.

rusu24edward commented:
> are any of them attempting to go the IBlanking route?

Yes, the ones that fail, like the example above, are triggering iblanking based on file size. I cannot look at the internals of the files because they are binary, but it seems clear from the result that the internal check is mislabeling them. Alternatively, they do have iblank info and we just haven't seen it until now because we didn't read it.

markcmiller86 commented on src/databases/PLOT3D/vtkPLOT3DReader.C:
```diff
@@ -790,20 +823,77 @@ vtkPLOT3DReader::ReadGrid(FILE *xyzFp)
+      }
+      else if (ib[i] == 1)
+      {
+        ib[i] = avtGhostNodeTypes::DUPLICATED_NODE;
```
We're overwritting VTK's `ib` data with VisIt ghost node type data here, right?

In the else block here, it doesn't make sense IMHO to assign VisIt's `avtGhostNodeType::DUPLICATED_NODE`. Instead, I think it should just clear `ib[i]` to zero `0` here as in no special ghost type is assigned for node `i`.

biagas replied:
> 
> 
> We're overwritting VTK's `ib` data with VisIt ghost node type data here, right?
> 
> In the else block here, it doesn't make sense IMHO to assign VisIt's `avtGhostNodeType::DUPLICATED_NODE`. Instead, I think it should just clear `ib[i]` to zero `0` here as in no special ghost type is assigned for node `i`.

@markcmiller86 I was confused by the fact that there isn't a value in avtGhostTypes to represent 'not-a-ghost' either for node or zone ghosts. Setting a node's ghost type to 0 is the same as setting it to DUPLICATED_NODE, codewise, is it not?

markcmiller86 replied:
Argh! Indeed `DUPLICATED_NODE` has value zero and that means `avtGhostTypes` left shifts 0 bits to assign the value for `ib[i]`.

I think `avtGhostTypes` is poorly designed in this regard then because it means it is not possible to have a node or a zone that has no *ghost* designation. Every node or zone is some kind of ghost. In particular, a zero'd ghost node array then means that all nodes are *duplicated*. Likewise for a ghost zone array. Does that make conceptual sense to anyone? It sure doesn't to me. It seems misleading.

`avtGhostTypes` desgin aside, using the `avtGhostType` symbol `DUPLICATED_NODE` in this code block in `vtkPLOT3DReader.C` IMHO repeats that confusion. Most readers could read `0` there and probably conclude *oh, ok, that branch is for nodes that are not ghosts* whereas now, readers are going to conclude that branch is for a different kind of ghost. In particular, a duplicated node.

You can only conclude, as you did, that it turns out the designation `DUPLICATED_NODE` has no effect if they go read the implementation of `avtGhostData` and discover

https://github.com/visit-dav/visit/blob/dfc29a1457ea9fc330befa7fc0b21ba1705946f7/src/avt/Pipeline/Data/avtGhostData.h#L109-L112

which returns `true` only for non-zero data.

markcmiller86 replied:
Actually, I am totally reading that ghosting logic in `avtGhostData.h` incorrectly! And, we appear to be using it incorrectly here and elsewhere too. 

`DUPLICATED_NODE` is a real ghost type and using it *should* result in a value of `0x01` in the ghost data value. That is because it is OR'ing in a `1` left shifted by `0` positions.

So, I rescind my comments regarding design of `avtGhostType` and `avtGhostData` and I'll amplify my point, don't use `DUPLICATED_NODE` there. Instead, assign 0.

markcmiller86 replied:
Oh, and we should never be *assigning* values of `avtGhostType::<SOME GHOST ENUM SYMBOL>`. We should be using the `AddGhostNodeType()` and `AddGhostZoneType()` functions there which handle the bit-shifting and or'ing.

brugger1 replied:
I agree with Mark, you should assign 0 if not a ghost node. If it is a ghost node, AddGhostNodeType() should be used. I also agree that this is very confusing.

rusu24edward replied:
Got it. I wasn't aware of the bit-shifting properties in ghost nodes/zones. I just saw the enumeration and hoped that duplicate node would be treated like a non-ghost-node since it is zero.

rusu24edward replied:
@brugger1 @markcmiller86 I've updated they way ghost nodes/zones are added. Please take a look and let me know if this is correct. I'm still seeing some weird behavior in the tests, which I'm trying to debug, and this may be the source of it if I've done this incorrectly.

brugger1 replied:
I think gnp will be uninitialized memory when it is created and since AddGhostNodeType() adds the bit shifted quantity to gnp[i], the result is undefined. I think you should zero gnp[i] before calling AddGhostNodeType(gnp[i],...).

biagas commented:
> 
> > are any of them attempting to go the IBlanking route?
> 
> Yes, the ones that fail, like the example above, are triggering iblanking based on file size. I cannot look at the internals of the files because they are binary, but it seems clear from the result that the internal check is mislabeling them. Alternatively, they do have iblank info and we just haven't seen it until now because we didn't read it.

I think it is probably the latter (they have iblank info), and we should turn off reading iblank to get the current results.
You could also add additional tests for these files to show the results of turning on iblank.

markcmiller86 commented on src/databases/PLOT3D/avtPLOT3DFileFormat.C:
```diff
@@ -210,9 +210,13 @@ avtPLOT3DFileFormat::avtPLOT3DFileFormat(const char *fname,
         {
             bool ib = readOpts->GetBool("IBlanking");
```
I would pick a slightly more explanatory string for this read option as in `Ignore Iblanking` or `Ignore iblanking if present`. Just the option `Iblanking` in the interface doesn't indicate what checking it on or off will really do.

rusu24edward commented:
> I think it is probably the latter (they have iblank info), and we should turn off reading iblank to get the current results.

If they do have iblanking in them, then wouldn't the correct results be the results we get from reading the file with iblanking?

biagas commented:
> 
> 
> > I think it is probably the latter (they have iblank info), and we should turn off reading iblank to get the current results.
> 
> If they do have iblanking in them, then wouldn't the correct results be the results we get from reading the file with iblanking?

Yes, there is a distinction between 'correct' and 'current' results.  My suggestion to create new tests was simply to have results for when iblanking is off and on.  Changing the current tests to ensure iblanking is off, preserves current baselines and preserves what we were actually testing before you enabled the iblanking functionality in the reader.

I've looked again at the test file.  It was definitely a poor choice on my part when I created the tests to keep the iblanking set to '1' (when testing read-options) when in fact it was being ignored.

Perhaps the best way forward, for those tests that aren't using read options: preserve the setting from the .vp3d file, which will mean rebaselining for those data files with IBLANKING turned on.

For those testing read options: turn off IBLANKING, which should preserve current baselines and demonstrate how one can override what is written in the .vp3d. file.

biagas commented:
All of this now has my head in a whirl.  Do users want a way to turn off iblanking?  If not, all my previous comments can be thrown out.

The 'IBLANKING' in the .vp3d file was meant to indicate whether or not the file contained IBLANKING, same with the Read option. 
It may have been created before the reader had a way of auto-checking for its existence. 

So now I'm thinking it should be used (current tests that specify the Read option 'IBLANKING' = 1 or have IBLANKING in the .vp3d should be rebaselined, because IBLANKING is in the file).

brugger1 commented:
Reading this discussion, I believe we should have 3 choices. Auto detect, On, Off. Auto detect is to automatically detect if iblanking is present and use it if it is. On is for manually telling it that iblanking is present and Off is for telling VisIt that iblanking is not present. It seems like On/Off are only for specifying manually if iblanking is present or not when the auto logic for some reason is getting it wrong. Descriptive settings would probably be a good idea, since this can be confusing.

rusu24edward committed:
IBlanking is an integer with 3 modes-auto, on, off

markcmiller86 commented:
FWIW, in other places we've adopted an "Always, Auto, Never" three-way logic model for similar things like scalable rendering, ignoring extents in Silo files, etc. Always means always utilize the feature. Never means never utilize it and Auto means let VisIt decide for you, perhaps with additional heurstic knobs such as in SR mode, the polygon count used in Auto mode.

rusu24edward committed:
Using enum instead of int for iblanking options

rusu24edward committed:
Settings copied to internal reader at requestinformation

rusu24edward commented:
> So now I'm thinking it should be used (current tests that specify the Read option 'IBLANKING' = 1 or have IBLANKING in the .vp3d should be rebaselined, because IBLANKING is in the file).

I like the idea of testing both on and off for the current tests. It's a bit of extra work, but I think it's the right way to go.

rusu24edward committed:
Using add ghost zone type

rusu24edward committed:
avtGhostNodes set correctly

biagas commented:
@markcmiller86 Do you still have logic in there to change the color if it matches the background?
Change background to red (first cycle color)
Open curve.visit, create a plot. The curve color isn't adjusted for the background, its red. And yes, I've verified I am working from the latest version of your branch this time.

Sorry, wrong PR!

rusu24edward committed:
Printing dataset at end of mesh read for debugging

rusu24edward committed:
Added process iblanking that is separate from internal iblanking so iblanking can be used to offset the grid without adding ghost nodes/zones

rusu24edward committed:
Fixed bool int typing

rusu24edward committed:
Process iblanking only on 1, not 2

rusu24edward committed:
Initialize ghost nodes with zeros

rusu24edward committed:
Setting for IBlanking in file and another for use if detected

rusu24edward committed:
Fixed PLO3D test for good viewing angles

rusu24edward committed:
Adjusted data and plot to correctly handle the usage of IBlanking data

rusu24edward committed:
Changed name back to IBlanking to not  change the python api

rusu24edward committed:
Merge branch 'bugfix/rusu1/visit-4632-plot3d-iblanking-support' of github.com:visit-dav/visit into bugfix/rusu1/visit-4632-plot3d-iblanking-support

rusu24edward committed:
Updated baselines for sslv because they have iblanking data for real

rusu24edward committed:
Updated baseline for wnb

rusu24edward committed:
Merge branch 'bugfix/rusu1/visit-4632-plot3d-iblanking-support' of github.com:visit-dav/visit into bugfix/rusu1/visit-4632-plot3d-iblanking-support

rusu24edward committed:
Plot3d test file using same api

brugger1 commented on src/doc/gui_manual/WorkingWithFiles/Read_Write_Options.rst:
```diff
@@ -110,6 +110,169 @@ all information related to the material configuration during the *open*. This wi
 lead to longer open times. A value of ``0`` here means to ignore any material
 information if present.
 
+PLOT3D
```
You should use "VisIt_" instead of "VisIt". You should have a single space between sentences. I noticed 2 spaces between sentences primarily in the example vp3d file.

brugger1 replied:
The above comments apply in general to the documentation.

brugger1 commented on src/databases/PLOT3D/vtkPLOT3DReader.C:
```diff
@@ -790,20 +855,85 @@ vtkPLOT3DReader::ReadGrid(FILE *xyzFp)
+    ghostZones->SetNumberOfValues(output->GetNumberOfCells());
+    ghostZones->SetName("avtGhostZones");
+    vtkIdList* ids = vtkIdList::New();
+    ids->SetNumberOfIds(8);
```
Setting the number of Ids is unnecessary here. The call to GetCellPoints will set it.

brugger1 commented on src/databases/PLOT3D/vtkPLOT3DReader.C:
```diff
@@ -790,20 +855,85 @@ vtkPLOT3DReader::ReadGrid(FILE *xyzFp)
+      unsigned char value = 0;
+      for (vtkIdType ptIdx = 0; ptIdx < numIds; ptIdx++)
+      {
+        if (gnp[ids->GetId(ptIdx)] == (1 << avtGhostNodeTypes::NODE_NOT_APPLICABLE_TO_PROBLEM))
```
Change this test to
avtGhostData::IsGhostZoneType(ids->GetId(ptIdx), NODE_NOT_APPLICABLE_TO_PROBLEM)

rusu24edward replied:
Thank you! Originally I had
```
avtGhostNodeTypes::NODE_NOT_APPLICABLE_TO_PROBLEM)
```
which did not work. Didn't see this function.

rusu24edward replied:
Just to confirm, I will use `IsGhostNodeType` instead of `IsGhostZoneType`.

brugger1 replied:
Oops, yes IsGhostNodeType is correct, not IsGhostZoneType.

brugger1 commented on src/databases/PLOT3D/vtkPLOT3DReader.C:
```diff
@@ -235,8 +266,13 @@ int vtkPLOT3DReader::AutoDetectionCheck(FILE* fp)
+    this->Internal->IBlanking = this->IBlankingInFile;
     }
+
+  std::cout << "After Auto-Detection, IBlankingInFile is " << this->IBlankingInFile << std::endl;
```
Remove this debugging information. It would probably be a good idea to write it to one of the debug logs.

brugger1 replied:
Oh, wait, this is a vtk function. Never mind.

markcmiller86 replied:
> Oh, wait, this is a vtk function. Never mind.

FWIW...I am not surre "...is a vtk function..." matters all that much for deciding whether debug log statements are allowed or a good idea. This is part of a reader's behavior and if it could help debugging, the more (debug statements) the merrier 😆 IMHO. But deciding which debug *level* is a good idea. Anything that would produce output that scales with `NumNodes` or `NumZones` should only go to level 5 logs (`debug5 << "After auto-detection...`).

rusu24edward replied:
I just used this to make sure that iblanking was being detected and applied as expected.

brugger1 commented on src/databases/PLOT3D/vtkPLOT3DReader.C:
```diff
@@ -171,22 +176,26 @@ int vtkPLOT3DReader::AutoDetectionCheck(FILE* fp)
-    }
+  }
 
+  std::cout << "Before Auto-Detection, IBlankingInFile is " << this->IBlankingInFile << std::endl;
```
More debugging information to remove.

brugger1 commented on src/databases/PLOT3D/avtPLOT3DFileFormat.C:
```diff
@@ -252,7 +262,8 @@ avtPLOT3DFileFormat::avtPLOT3DFileFormat(const char *fname,
         reader->BinaryFileOn();
         reader->SetByteOrderToBigEndian();
-        reader->IBlankingOff();
+        reader->SetIBlankingInFile(0); // Auto setting, TODO: Use enum
```
You have a bunch of TODO, use enum for SetIBlankingInFile.

rusu24edward committed:
Updated PLOT3d test

rusu24edward committed:
Merge branch 'bugfix/rusu1/visit-4632-plot3d-iblanking-support' of github.com:visit-dav/visit into bugfix/rusu1/visit-4632-plot3d-iblanking-support

rusu24edward committed:
Using enum in Options

rusu24edward committed:
Removed debugging print statements

rusu24edward committed:
Enhanced plot3d docs a bit

rusu24edward committed:
Fixed enum name for iblanking

rusu24edward committed:
Further updates to docs

rusu24edward committed:
Updated relnotes

rusu24edward committed:
Put new test into a function for easier commenting

rusu24edward committed:
Small changes to docs

rusu24edward committed:
Using IsGhostNodeType

brugger1 commented:
Oops, should be IsGhostNodeType not IsGhostZoneType.

rusu24edward committed:
Simplified wnb testing

rusu24edward committed:
Merge branch 'bugfix/rusu1/visit-4632-plot3d-iblanking-support' of github.com:visit-dav/visit into bugfix/rusu1/visit-4632-plot3d-iblanking-support

rusu24edward committed:
Fixed use of IsGhostNodeType

rusu24edward committed:
Merge branch 'bugfix/rusu1/visit-4632-plot3d-iblanking-support' of github.com:visit-dav/visit into bugfix/rusu1/visit-4632-plot3d-iblanking-support

rusu24edward committed:
Missing quote in docs

rusu24edward committed:
Don't set number of ids

rusu24edward committed:
Merge branch 'bugfix/rusu1/visit-4632-plot3d-iblanking-support' of github.com:visit-dav/visit into bugfix/rusu1/visit-4632-plot3d-iblanking-support

rusu24edward committed:
Removed vtkVisItUtility debugging

rusu24edward committed:
Merged develop and resolved conflict in relnotes

rusu24edward commented:
Big thanks to @biagas @brugger1 and @markcmiller86 for your detailed review so far. I believe I have addressed all the issues that were raised in the draft PR. I've converted to a normal PR and re-requested your reviews. Thanks!

biagas commented:
@rusu24edward I checked out your branch and compiled it on Windows.
I'm trying to understand the new iblanking options.

I'm looking at the WingBodyTail dataset, telling visit to open the .vp3d file
First using the 'guess from filename' option.
I see a big difference from 3.1.3 drawing, so I know that Iblanking and ghosting has been applied, I also used -dump and saw ghost information in the vtk files.

Then I tried opening it using Read options and turning off iblanking (IBLANKING set to NEVER and 'Use IBlanking if present' unchecked), the ghosting information is still created (it is still in the dumped vtk files).

Am I incorrect in thinking that the read options I used should have told VisIt to ignore the iblanking, and thus not create any ghost information?

rusu24edward commented:
@biagas I'm not sure how the ghosting information is getting into the vtk files. In order for ghost information to be created and added to the dataset, we must satisfy [this condition](https://github.com/visit-dav/visit/blob/bugfix/rusu1/visit-4632-plot3d-iblanking-support/src/databases/PLOT3D/vtkPLOT3DReader.C#L858), which is that we need both IBlanking in the file and also we need to Use IBlanking if Present. I double checked this with some debugging statements just now. Does ghost information get created when opening this file on develop?

biagas commented:
> Does ghost information get created when opening this file on develop?

No ghosting information coming from the reader on develop.

brugger1 reviewed (dismissed):
Looks good. I tested with SpaceShuttleLaunchVehicle and it behaved like I expected for the various combinations of i blanking options. I was surprised how crude the iblanking was at removing cells to expose the space shuttle, but I believe the images were correct.

rusu24edward committed:
Fixed misleading comment

rusu24edward commented:
Todo: Update docs to be clear that vp3d file overwrites whatever the read options say

biagas reviewed (approved):
Great work!

rusu24edward committed:
Merge branch 'develop' into bugfix/rusu1/visit-4632-plot3d-iblanking-support
